scripts/airgapped/get-all-images.py

```diff
@@ -0,0 +1,177 @@
+
+    logging.info(f"Cloning repo {repo_url}")
+    repo = git.Repo.clone_from(repo_url, str(repo_name))
+
+    logging.info(f"Checking out to branch {branch}")
+    repo.git.checkout(branch)
+
+    return repo_name
+
+
+def get_analytics_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a charm developed by us, by:
+    1. Cloning the repo of the charm
+    2. Running the repo's tools/get-images.sh
+    3. Delete the repo
+
+    If the tools/get-images.sh of a repo fails for any reason then this
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
+
+    logging.info(f"Executing repo's {GET_IMAGES_SH} script")
+    process = subprocess.Popen(["bash", "tools/get-images.sh"], cwd=repo_dir,
+                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = process.communicate()
+    if stderr:
+        raise ValueError("Script '%s' for charm '%s' had error logs: \n%s" %
+                         (GET_IMAGES_SH, app["charm"], stderr.decode("utf-8"))
+                         )
+    images = stdout.decode("utf-8").split("\n")
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_bundle_images(bundle_dict: dict) -> list[str]:
+    """Return a list of images used by a bundle"""
+    images = []
+
+    for app_name, app in bundle_dict["applications"].items():
+        logging.info(f"Handling app {app_name}")
+
+        # exclude repos we know don't have images
+        if app_name in EXCLUDE_REPOS:
+            logging.info("Ignoring charm %s", app["charm"])
+            continue
+
+        # Follow default image-gather process for dependency apps
+        if is_dependency_app(app):
+            logging.info("Dependency app '%s' with charm '%s'", app_name,
+                         app["charm"])
+            images.extend(get_dependency_app_images(app))
+            continue
+
+        # Ensure analytics app has necessary github repo/branch metadata
+        if not bundle_app_contains_gh_metadata(app):
+            raise KeyError(
```

Comment:
I'd like the script to exit whenever something unexpected happens, so that it always fails very vocally to ensure we will catch it.

Would you suggest we would log a message and terminate the script in a different way here?

Reply:
I actually think we are not waiting for something unexpected to happen, we are very clearly saying "if the app does not have GH metadata, we should raise", so for starters, we know we want to interrupt the execution whenever there is a missing GH metadata. 

I agree that the script should be interrupted and tell the user what actually happened, but I believe just raising won't do it gracefully. It will be very verbose, that is right, but that's because the execution will be interrupted and the user will be shown a traceback with `KeyError` that says "Application '%s' doesn't include expected gh metadata keys" %". 

> Would you suggest we would log a message and terminate the script in a different way here?

I see two options for approaching this better:

1. As I mentioned before: save a list of images that did not have the right GH metadata, log this saying "Skipping images for charm X because it does not have the required metadata", and continue with the rest of the applications in the bundle?
2. (I like this approach better) Handle the exception, log, and exit with an error code. For that I would:

- Merge the `is_dependency_app()` and `bundle_app_contains_gh_metadata()` methods into one that validates the metadata, perhaps `validate_gh_metadata_in_bundle(bundle_path: Path)`:

```python3
import sys
from pathlib import Path

import yaml
from jsonschema import validate
from jsonschema.exceptions import ValidationError

bundle_schema = {} # define the bundle json schema here with the required fields "_gh_repo_name" and "_gh_branch_name"

def validate_gh_metadata_in_bundle(bundle_path: str) -> None:
  """
    Validate the bundle_path against a JSON schema with required fields for GH metadata.
    
    Args:
       bundle_path(str): path to the bundle file as a string
    # See the note below for raising
    Raises:
        ValidationError: if the required GH metadata is not present in the bundle
  """
  try:
      validate(yaml.safe_load(Path(bundle_path).read_text()), bundle_schema)
  except ValidationError as e:
      log.error(f"Issue with validation: {e.message}") # <-- this will show a message like: "Issue with validation: '_gh_repo_name' is a required property
      # Option A: raise and let the exception gets handled somewhere else (perhaps in main)
      raise e # <-- we could re-raise the exception here to be handled elsewhere OR log and sys.exit(1) here
      # Option B: do not raise, just exit with error code
      sys.exit(1)
```

- Once I have a validator, I can call it as the first thin in my `main` method, as I want everything to be blocked until the bundle has the right schema.

Also, if you still think we should go with your approach, do you think [`KeyError`](https://docs.python.org/3/library/exceptions.html#KeyError) fits the use case?

Reply:
I've updated the code to instead of raising the KeyError to `log.error` and `sys.exit(1)`. If in the future this script will be used by another script and we'd really care about capturing this exception but continuing, but until then I agree we can just exit the script.

I didn't go the route of doing the jsonschema as it seems a bit overkill to define the relations and validate the bundle for this case. But let's re-consider if we feel the validation of the yaml gets more and more involved.

Reply:
> I didn't go the route of doing the jsonschema as it seems a bit overkill to define the relations and validate the bundle for this case. But let's re-consider if we feel the validation of the yaml gets more and more involved.

What do you mean define the relations and validate the bundle?

The validation I am proposing is for making sure that the GH metadata is there as we want it; this would replace the `validate_gh_metadata_in_bundle(bundle_path: Path)` method with an approach that validates the schema of the yaml bundle file in a more yaml/json-oriented approach instead of iterating over all the charm objects.

It does not look overkill to me, I'd say it's a quick check that `jsonschema` will do for you, and this can save you execution time as if the validation does not succeeds, the script will be interrupted; contrary to parsing the bundle file, get the charm objects as dictionaries, and look for the same string in the keys of the dict.

From the working example I provided, you would just need to define the `bundle_schema` with the required fields (the two fields we care about), other than that there is not much else to do.

Reply:
> this can save you execution time as if the validation does not succeeds, the script will be interrupted; contrary to parsing the bundle file, get the charm objects as dictionaries, and look for the same string in the keys of the dict

This is a very good point! I've modified the logic to first validate the whole bundle before continuing to running the scripts.

On the jsonschema, it still looks overkill and that it will make the code more complex right now for small gain. I've created a follow-up issue though to discuss this further https://github.com/canonical/bundle-kubeflow/issues/1059

So I'll go on and resolve this discussion